### PR TITLE
MueLu: Refactor CoalesceDropFactory_kokkos

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu 

## Motivation
Refactor CoalesceDropFactory_kokkos and break it up into smaller, composable functors.

### CODE REVIEW

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_BoundaryDetection.hpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  using scalar_type        = typename local_matrix_type::value_type;
+  using local_ordinal_type = typename local_matrix_type::ordinal_type;
+  using memory_space       = typename local_matrix_type::memory_space;
+
+  local_matrix_type A;
+  functor_type_0 functor0;
+  functor_type_1 functor1;
+  functor_type_2 functor2;
+  functor_type_3 functor3;
+
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_BoundaryDetection.hpp:127]: document assumptions on inputs, and how composition works

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_BoundaryDetection.hpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        }
+      }
+      if constexpr (useGreedyDirichlet) {
+        if (rowIsDirichlet) {
+          boundaryNodes(rblid) = true;
+          return;
+        }
+      } else {
+        if (!rowIsDirichlet) {
+          boundaryNodes(rblid) = false;
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_BoundaryDetection.hpp:229]: MAybe remove?

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_ClassicalDropping.hpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  void operator()(const local_ordinal_type rlid) const {
+    auto row      = A.rowConst(rlid);
+    size_t offset = A.graph.row_map(rlid);
+    for (local_ordinal_type k = 0; k < row.length; ++k) {
+      auto clid = row.colidx(k);
+
+      auto val    = row.value(k);
+      auto aiiajj = ATS::magnitude(diag(rlid)) * ATS::magnitude(diag(clid));  // |a_ii|*|a_jj|
+      auto aij2   = ATS::magnitude(val) * ATS::magnitude(val);                // |a_ij|^2
+
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_ClassicalDropping.hpp:64]: document how this works

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_ClassicalDropping.hpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  using diag_view_type = typename Kokkos::DualView<const scalar_type*, Kokkos::LayoutStride, typename Node::device_type, Kokkos::MemoryUnmanaged>::t_dev;
+
+  local_matrix_type A;
+  Teuchos::RCP<diag_vec_type> diagVec;
+  diag_view_type diag;  // corresponds to overlapped diagonal
+  magnitudeType eps;
+  results_view results;
+
+ public:
+  SignedClassicalRSDropFunctor(matrix_type& A_, magnitudeType threshold, results_view& results_)
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_ClassicalDropping.hpp:96]: COme up with names

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_ClassicalDropping.hpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      auto clid = row.colidx(k);
+
+      auto val                  = row.value(k);
+      auto aiiajj               = ATS::magnitude(diag(rlid)) * ATS::magnitude(diag(clid));  // |a_ii|*|a_jj|
+      const bool is_nonpositive = ATS::real(val) <= mATS::zero();
+      magnitudeType aij2        = ATS::magnitude(val) * ATS::magnitude(val);  // |a_ij|^2
+      // + |a_ij|^2, if a_ij < 0, - |a_ij|^2 if a_ij >=0
+      if (is_nonpositive)
+        aij2 = -aij2;
+      results(offset + k) = Kokkos::max((aij2 <= eps * eps * aiiajj) ? DROP : KEEP,
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_ClassicalDropping.hpp:168]: wonky

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_kokkos_def.hpp [hunk 6]
```diff
@@ -23,435 +24,682 @@
...
+  // - BoundaryDetection::RowSumFunctor
+  //   Marks rows as Dirichlet bases on row-sum criterion
+
+  // Dirichlet nodes
+  auto boundaryNodes = boundary_nodes_type("boundaryNodes", lclA.numRows());  // initialized to false
+  {
+    SubFactoryMonitor mBoundary(*this, "Boundary detection", currentLevel);
+
+    // macro that applies boundary detection functors
+#define runBoundaryFunctor(...)                                                 \
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_kokkos_def.hpp:260]: namespace

File: packages/muelu/test/unit_tests/ParameterList/ParameterListInterpreter.cpp [hunk 7]
```diff
@@ -82,7 +82,12 @@ TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL(ParameterListInterpreter, BlockCrs, Scalar, Lo
       if (found == std::string::npos) continue;
 
       out << "Processing file: " << fileList[i] << std::endl;
-      ParameterListInterpreter mueluFactory("ParameterList/ParameterListInterpreter/" + fileList[i], *comm);
+
+      Teuchos::RCP<Teuchos::ParameterList> mueluList = rcp(new Teuchos::ParameterList());
+      Teuchos::updateParametersFromXmlFileAndBroadcast("ParameterList/ParameterListInterpreter/" + fileList[i], mueluList.ptr(), *comm);
+      mueluList->set("use kokkos refactor", false);
+
+      ParameterListInterpreter mueluFactory(*mueluList, comm);
```
Comment (cgcgcg) [inline packages/muelu/test/unit_tests/ParameterList/ParameterListInterpreter.cpp:90]: Why?

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_MatrixConstruction.hpp [hunk 8]
```diff
@@ -0,0 +1,1029 @@
...
+    scalar_type diagCorrection      = zero;
+    local_ordinal_type diagOffset   = -1;
+    for (local_ordinal_type k = 0; k < rowA.length; ++k) {
+      if constexpr (lumping) {
+        local_ordinal_type clid = rowA.colidx(k);
+        if (rlid == clid) {
+          diagOffset = j;
+        }
+      }
+      if (results(K + k) == KEEP) {
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_MatrixConstruction.hpp:404]: K and k

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_DroppingCommon.hpp [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+      if ((results(offset + k) == KEEP) && (rlid != clid))
+        return;
+    }
+    boundaryNodes(rlid) = true;
+    for (local_ordinal_type k = 0; k < row.length; ++k) {
+      auto clid = row.colidx(k);
+      if (rlid == clid)
+        results(offset + k) = KEEP;
+      else
+        results(offset + k) = BOUNDARY;
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_DroppingCommon.hpp:204]: What's the diff between DROP and BOUNDARY?

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_CutDrop.hpp [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+    if (start > 0)
+      start = start - 1;
+    else {
+      end       = end - 1;
+      auto temp = v(0);
+      v(0)      = v(end);
+      v(end)    = temp;
+    }
+    size_t root = start;
+    while (2 * root + 1 < end) {
```
Comment (cgcgcg) [inline packages/muelu/src/Graph/MatrixTransformation/MueLu_CutDrop.hpp:428]: @jhux2

Comment (jhux2) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CutDrop.hpp:428]: My concern was related to shell sort, not heap sort.  Question:  is a stable sort needed here?

Comment (cgcgcg) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_CutDrop.hpp:428]: I don't think so? If two entries give the same value for sorting then either both should be kept or dropped.

File: packages/muelu/src/Graph/MatrixTransformation/MueLu_BoundaryDetection.hpp (hunk 2 above)
Comment (cgcgcg) [reply packages/muelu/src/Graph/MatrixTransformation/MueLu_BoundaryDetection.hpp:132]: `boundaryNodes` is initialized to all `false`. Then we apply one or more functors that can switch entries to `true`. We probably don't want to override previous decisions to mark a node as Dirichlet. Currently we don't use more than one functor, but that seems reasonable as a design decision, right?

**Review by jhux2 — APPROVED** [review]
Lgtm.  @cgcgcg and I went through this together.

**Pull request merged**